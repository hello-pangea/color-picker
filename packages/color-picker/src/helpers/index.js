export * from "./checkboard";
export * from "./color";
