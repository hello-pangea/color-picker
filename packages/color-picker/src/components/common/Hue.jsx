import React, { Component, PureComponent } from "react";
import * as hue from "../../helpers/hue";

export class Hue extends (PureComponent || Component) {
  componentWillUnmount() {
    this.unbindEventListeners();
  }

  handleChange = (e) => {
    const change = hue.calculateChange(
      e,
      this.props.direction,
      this.props.hsl,
      this.container
    );
    change &&
      typeof this.props.onChange === "function" &&
      this.props.onChange(change, e);
  };

  handleMouseDown = (e) => {
    this.handleChange(e);
    window.addEventListener("mousemove", this.handleChange);
    window.addEventListener("mouseup", this.handleMouseUp);
  };

  handleMouseUp = () => {
    this.unbindEventListeners();
  };

  unbindEventListeners() {
    window.removeEventListener("mousemove", this.handleChange);
    window.removeEventListener("mouseup", this.handleMouseUp);
  }

  render() {
    const { direction = "horizontal" } = this.props;

    const styles = {
      hue: {
        position: "absolute",
        inset: "0px",
        borderRadius: this.props.radius,
        boxShadow: this.props.shadow,
      },
      container: {
        padding: "0 2px",
        position: "relative",
        height: "100%",
        borderRadius: this.props.radius,
      },
      pointer: {
        position: "absolute",
        left:
          direction === "vertical"
            ? "0px"
            : `${(this.props.hsl.h * 100) / 360}%`,
        top:
          direction === "vertical"
            ? `${-((this.props.hsl.h * 100) / 360) + 100}%`
            : undefined,
      },
      slider: {
        marginTop: "1px",
        width: "4px",
        borderRadius: "1px",
        height: "8px",
        boxShadow: "0 0 2px rgba(0, 0, 0, .6)",
        background: "#fff",
        transform: "translateX(-2px)",
      },
    };

    return (
      <div style={styles.hue}>
        <div
          className={`hue-${direction}`}
          style={styles.container}
          ref={(container) => (this.container = container)}
          onMouseDown={this.handleMouseDown}
          onTouchMove={this.handleChange}
          onTouchStart={this.handleChange}
        >
          <style>{`
            .hue-horizontal {
              background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }

            .hue-vertical {
              background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }
          `}</style>
          <div style={styles.pointer}>
            {this.props.pointer ? (
              <this.props.pointer {...this.props} />
            ) : (
              <div style={styles.slider} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default Hue;
