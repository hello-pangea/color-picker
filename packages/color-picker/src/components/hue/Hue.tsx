import merge from "lodash/merge";
import React from "react";
import { useColor, withColorProvider } from "../../context/useColor";
import { ChangeColor } from "../../types/colors";
import { Hue } from "../common";
import HuePointer from "./HuePointer";

export type HuePickerProps = {
  width?: string | number;
  height?: string | number;
  direction?: "horizontal" | "vertical";
  pointer: typeof HuePointer;
  className?: string;
  styles?: Record<string, React.CSSProperties>;
};

export function HuePicker({
  width = "316px",
  height = "16px",
  direction = "horizontal",
  pointer = HuePointer,
  styles: passedStyles = {},
  className = "",
}: HuePickerProps) {
  const { colors, changeColor } = useColor();
  const { hsl } = colors;

  const styles = merge<
    Record<string, React.CSSProperties>,
    Record<string, React.CSSProperties>
  >(
    {
      picker: {
        position: "relative",
        width,
        height,
      },
      hue: {
        borderRadius: "2px",
      },
    },
    passedStyles
  );

  // Overwrite to provide pure hue color
  const handleChange = (data: ChangeColor) =>
    changeColor({
      a: 1,
      h: typeof data !== "string" && "h" in data ? data.h : 0,
      l: 0.5,
      s: 1,
    });

  return (
    <div style={styles.picker} className={`hue-picker ${className}`}>
      <Hue
        {...styles.hue}
        hsl={hsl}
        pointer={pointer}
        onChange={handleChange}
        direction={direction}
      />
    </div>
  );
}

export default withColorProvider(HuePicker);
