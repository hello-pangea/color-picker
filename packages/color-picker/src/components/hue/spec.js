/* global test, expect */

import React from "react";
import renderer from "react-test-renderer";
import { red } from "../../helpers/color";

import Hue from "./Hue";
import HuePointer from "./HuePointer";

test("Hue renders correctly", () => {
  const tree = renderer.create(<Hue {...red} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Hue renders vertically", () => {
  const tree = renderer
    .create(<Hue {...red} width={20} height={200} direction="vertical" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Hue renders custom styles correctly", () => {
  const tree = renderer
    .create(
      <Hue
        {...red}
        styles={{ default: { picker: { boxShadow: "0 0 10px red" } } }}
      />
    )
    .toJSON();
  expect(tree.props.style.boxShadow).toBe("0 0 10px red");
});

test("HuePointer renders correctly", () => {
  const tree = renderer.create(<HuePointer />).toJSON();
  expect(tree).toMatchSnapshot();
});
