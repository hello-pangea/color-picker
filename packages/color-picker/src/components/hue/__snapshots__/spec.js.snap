// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hue renders correctly 1`] = `
<div
  className="hue-picker "
  style={
    {
      "height": "16px",
      "position": "relative",
      "width": "316px",
    }
  }
>
  <div
    style={
      {
        "borderRadius": undefined,
        "boxShadow": undefined,
        "inset": "0px",
        "position": "absolute",
      }
    }
  >
    <div
      className="hue-horizontal"
      onMouseDown={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        {
          "borderRadius": undefined,
          "height": "100%",
          "padding": "0 2px",
          "position": "relative",
        }
      }
    >
      <style>
        
            .hue-horizontal {
              background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }

            .hue-vertical {
              background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }
          
      </style>
      <div
        style={
          {
            "left": "69.44444444444443%",
            "position": "absolute",
            "top": undefined,
          }
        }
      >
        <div
          style={
            {
              "backgroundColor": "rgb(248, 248, 248)",
              "borderRadius": "50%",
              "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "height": "18px",
              "transform": "translate(-9px, -1px)",
              "width": "18px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Hue renders vertically 1`] = `
<div
  className="hue-picker "
  style={
    {
      "height": 200,
      "position": "relative",
      "width": 20,
    }
  }
>
  <div
    style={
      {
        "borderRadius": undefined,
        "boxShadow": undefined,
        "inset": "0px",
        "position": "absolute",
      }
    }
  >
    <div
      className="hue-vertical"
      onMouseDown={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        {
          "borderRadius": undefined,
          "height": "100%",
          "padding": "0 2px",
          "position": "relative",
        }
      }
    >
      <style>
        
            .hue-horizontal {
              background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
                33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to right, #f00 0%, #ff0
                17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }

            .hue-vertical {
              background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
                #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
              background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
                #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            }
          
      </style>
      <div
        style={
          {
            "left": "0px",
            "position": "absolute",
            "top": "30.55555555555557%",
          }
        }
      >
        <div
          style={
            {
              "backgroundColor": "rgb(248, 248, 248)",
              "borderRadius": "50%",
              "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
              "height": "18px",
              "transform": "translate(-3px, -9px)",
              "width": "18px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`HuePointer renders correctly 1`] = `
<div
  style={
    {
      "backgroundColor": "rgb(248, 248, 248)",
      "borderRadius": "50%",
      "boxShadow": "0 1px 4px 0 rgba(0, 0, 0, 0.37)",
      "height": "18px",
      "transform": "translate(-9px, -1px)",
      "width": "18px",
    }
  }
/>
`;
